src/bits/mod.rs

```diff
@@ -53,3 +54,23 @@ impl<T> BooleanRefOps for T where
     T: for<'a> BitXor<&'a Self, Output = Self> + for<'a> BitXorAssign<&'a Self>
 {
 }
+
+/// Trait for items that have fixed-byte length representation.
+pub trait Serializable: Sized {
+    /// Required number of bytes to store this message on disk/network
+    const SIZE_IN_BYTES: usize;
+
+    /// Serialize this message to a mutable slice. Implementations need to ensure `buf` has enough
+    /// capacity to store this message.
+    ///
+    /// ## Errors
+    /// Returns an error if `buf` does not have enough capacity to store at least `SIZE_IN_BYTES` more
+    /// data.
+    fn serialize(self, buf: &mut [u8]) -> io::Result<()>;
```

Comment:
Do we really want to return `Result` from these?  It seems that we might instead insist that the caller provide a precisely-sized slice to both functions.  After all, the caller knows.

I see a lot of checking in the implementations of these and that is really unnecessary.  Given that these are likely on hot paths, that extra checking comes with costs.

(Ideally, this would be a runtime constraint and the argument would be `&[u8; Self::SIZE_IN_BYTES]`, but I don't know if that works, even with generic const expressions.)